File: Data Structures & Algorithms/remove-node-from-end-of-linked-list/submission-4.js
/**
 * Definition for singly-linked list.
 * class ListNode {
 *     constructor(val = 0, next = null) {
 *         this.val = val;
 *         this.next = next;
 *     }
 * }
 */

class Solution {
    /**
     * @param {ListNode} head
     * @param {number} n
     * @return {ListNode}
     */
    removeNthFromEnd(head, n) {

        let length = 0
        let current = head

        while (current){
            length += 1
            current = current.next
        }
        if (n == length){
            return head.next
        }
        current = head;
        for (let i = 0; i < length - n - 1; i++){
            current = current.next
        }
        current.next = current.next.next

        return head
    }
}

File: Data Structures & Algorithms/reorder-linked-list/submission-4.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next

class Solution:
    def reorderList(self, head: Optional[ListNode]) -> None:

        if not head and not head.next:
            return head

        slow = fast = head
        while fast and fast.next:
            fast = fast.next.next
            slow = slow.next


        prev = None
        curr =  slow
        while curr:
            next_ = curr.next
            curr.next = prev
            prev =  curr
            curr = next_

        first = head
        second = prev
        while second.next:
            temp1 = first.next
            first.next  = second
            first = temp1

            temp2 = second.next
            second.next = first
            second = temp2

        
        